### DISCUSSION TITLE
Trimming, ReflectionBindingExtension and publish warnings

### DISCUSSION THREAD

**Discussion by wgraaf** [AvaloniaUI/Avalonia#16791] [Q&A]
I have an Avalonia application that I publish with the 'Produce single file' and 'Trim unused code'.
This results in hundreds of warnings like:

> Using member 'Avalonia.Markup.Xaml.MarkupExtensions.ReflectionBindingExtension.ReflectionBindingExtension(String)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. BindingExpression and ReflectionBinding heavily use reflection. Consider using CompiledBindings instead.
 
The published application works without issues, as far as I can tell.
If I unselect 'Trim unused code' the warnings disappear, but the executable size doubles to ~90MB, which is undesired.

Is it safe to ignore the warnings?
What kind of issues should I be expecting?

**Comment by stevemonaco** [comment]
The quick answer is: stop using reflection bindings. There are very few legitimate uses for them that compiled bindings do not cover. Apps with trimming warnings will work until they don't. You don't have any guarantees because you've violated the analysis. You can mitigate by making your entire assembly non-trimmable (ie. the project with the Avalonia views) by following advice [here](https://github.com/dotnet/runtime/discussions/95891#discussioncomment-7829505).

**Reply by wgraaf** [reply]
I created a project with just an empty window, specifying , x:CompileBindings="True", and I am STILL getting trim warnings when publishing  self-contained to single file with trimming. How can I get rid of these?

Assembly 'Avalonia.Base' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries	AvTest	C:\Users\wgraaf\.nuget\packages\avalonia\11.1.3\lib\net6.0\Avalonia.Base.dll	0		
AvTest.ViewLocator.Build(Object): Unrecognized value passed to the parameter 'typeName' of method 'System.Type.GetType(String)'. It's not possible to guarantee the availability of the target type.	AvTest	C:\AvTest\AvTest\ViewLocator.cs	15		
Unrecognized value passed to the parameter 'typeName' of method 'System.Type.GetType(String)'. It's not possible to guarantee the availability of the target type.	AvTest	C:\AvTest\AvTest\ViewLocator.cs	15		
Assembly 'Avalonia.Win32' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries	AvTest	C:\Users\wgraaf\.nuget\packages\avalonia.win32\11.1.3\lib\net6.0\Avalonia.Win32.dll	0		
Assembly 'Avalonia.DesignerSupport' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries	AvTest	C:\Users\wgraaf\.nuget\packages\avalonia\11.1.3\lib\net6.0\Avalonia.DesignerSupport.dll	0		
Assembly 'DynamicData' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries	AvTest	C:\Users\wgraaf\.nuget\packages\dynamicdata\7.9.5\lib\net6.0\DynamicData.dll	0		
Assembly 'ReactiveUI' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries	AvTest	C:\Users\wgraaf\.nuget\packages\reactiveui\18.3.1\lib\net6.0\ReactiveUI.dll	0		
Assembly 'Splat' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries	AvTest	C:\Users\wgraaf\.nuget\packages\splat\14.4.1\lib\net6.0\Splat.dll	0		
Assembly 'System.Reactive' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries	AvTest	C:\Users\wgraaf\.nuget\packages\system.reactive\5.0.0\lib\net5.0\System.Reactive.dll	0		
Assembly 'Tmds.DBus.Protocol' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries	AvTest	C:\Users\wgraaf\.nuget\packages\tmds.dbus.protocol\0.16.0\lib\net6.0\Tmds.DBus.Protocol.dll	0		
Avalonia.Data.Core.TargetTypeConverter.DefaultConverter.TryConvert(Object, Type, CultureInfo, Object&): Using member 'System.ComponentModel.TypeDescriptor.GetConverter(Type)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Generic TypeConverters may require the generic types to be annotated. For example, NullableConverter requires the underlying type to be DynamicallyAccessedMembers All.	AvTest	ILLink	0		
Avalonia.Data.Core.TargetTypeConverter.DefaultConverter.TryConvert(Object, Type, CultureInfo, Object&): Using member 'System.ComponentModel.TypeDescriptor.GetConverter(Type)' which has 'RequiresUnreferencedCodeAttribute' can break functionality when trimming application code. Generic TypeConverters may require the generic types to be annotated. For example, NullableConverter requires the underlying type to be DynamicallyAccessedMembers All.	AvTest	ILLink	0		
Could not resolve assembly 'Avalonia.Themes.Fluent'.	AvTest	Roots.xml	4

**Reply by maxkatz6** [reply]
> ViewLocator.cs

That's in your code. Don't use reflection in the view locator if you want AOT/trimmability.

> Avalonia.Win32
> Tmds.DBus.Protocol

These warnings were resolved in 11.2 beta. Weren't critical before either.

> DynamicData
> ReactiveUI
> Splat
> System.Reactive

That's not Avalonia. These might or might not work.

> Avalonia.DesignerSupport

This one shouldn't really be used in release builds, it's only used when run from the designer. Weird you have that.

**Reply by maxkatz6** [reply]
To add on reactive libraries - most of the APIs there should work fine with AOT, but you have to live with these warnings for now.
You can try updating to the latest versions of these packages, it might be better there.
Either way, you can contact these repositories on github directly for advice with AOT. 

Avalonia also works with other MVVM frameworks, like Community Toolkit.
